python/drned_xmnr/op/filtering.py

```diff
@@ -356,184 +355,312 @@ def event_generator(consumer):
-        if event.action == 'commit':
-            return (event.produce_line(), DrnedCommitState())
-        elif event.action == 'compare_config':
-            return (event.produce_line(), DrnedCompareState())
-        else:
-            return event.produce_line()
+        if isinstance(event, StartState) or isinstance(event, Transition):
+            return (True,
+                    [GenS(DrnedPrepare), TransitionS(),
+                     GenS(TransFailed), GenS(InitFailed),
+                     self],
+                    event.produce_line())
+        if isinstance(event, DrnedTeardown):
+            return (True, [ActionS('restore')], event.produce_line())
+        return (False, [])
 
 
-class DrnedCommitState(LogState):
+class GenS(LS):
+    '''A generic log state.
+
+    Every event of given type is accepted (and its line is produced),
+    everything else is left for further processing.
+
+    (So e.g. GenS(prepare) -> prepare | [])
+
+    '''
+
+    def __init__(self, eventclass):
+        self.eventclass = eventclass
+        self.name = 'gen({})'.format(eventclass.__name__)
+
     def handle(self, event):
-        if isinstance(event, DrnedCommitLogEvent):
-            event.mark_complete()
-            return DrnedCommitResultState()
-        elif isinstance(event, DrnedCommitNNEvent):
-            event.mark_complete()
-            return DrnedCommitNNState()
-        elif isinstance(event, DrnedEmptyCommit):
-            event.mark_complete()
-            return (event.produce_line(), TERMINATED)
-        elif isinstance(event, DrnedCommitEvent):
-            event.mark_complete()
-            return TERMINATED
-        return TERMINATED
-
-
-class DrnedCommitResultState(LogState):
+        if isinstance(event, self.eventclass):
+            return (True, [], event.produce_line())
+        return (False, [])
+
+
+class WalkTransitionsS(LS):
+    '''
+    WT -> (pytest Transition)* (teardown Restore Commit Compare)?
+    '''
+
+    name = 'transitions'
+
     def handle(self, event):
-        if isinstance(event, DrnedCommitResult):
-            event.mark_complete()
-            if event.success:
-                return (event.produce_line(), TERMINATED)
-            else:
-                return DrnedFailureState()
-        return TERMINATED
+        if isinstance(event, PyTest):
+            return (True, [TransitionS(), self], event.produce_line())
+        if isinstance(event, DrnedTeardown):
+            return (True,
+                    [GenS(DrnedRestore),
+                     ActionS('commit', [CommitS()]),
+                     ActionS('compare_config', [CompareS()])], event.produce_line())
+
+        return (False, [])
+
 
+class SilentTransitionsS(LS):
+    'Produces output for the set of event types, ignores others.'
+
+    name = 'silent-transitions'
+    evtclasses = {InitStates, StartState, Transition,
+                  InitFailed, TransFailed,
+                  PyTest, DrnedTeardown}
+
```

Comment:
When allowing walk-states to recover from errors and continue I think it make sense to add DrnedCompareEvent to this list, I think.

Reply:
I don't really agree, "compare" is on pretty much the same level as "commit", and the latter is not included.

Reply:
But Compare is only shown when an error happen, it matches the diff output. When walk-states bail out at the first error it doesn't really matter but when we recover and continue I think it makes sense to indicate state errors in the log.

Reply:
I'm not sure I understand; compare is shown for every transition in the "drned-overview" log detail. I believe for walk failures we should add a new event and add it here (and I believe ialso to the Transition state handler).